probe-rs/src/probe/stlink/mod.rs

```diff
@@ -834,6 +1101,632 @@ impl From<StlinkError> for ProbeCreationError {
+
+        let mut buff = [0u8];
+        self.read_8(address, &mut buff)?;
+
+        Ok(buff[0])
+    }
+
+    fn read_32(&mut self, address: u32, data: &mut [u32]) -> Result<(), ProbeRsError> {
+        self.probe.select_ap(self.access_port)?;
+
+        let received_words = self.probe.probe.read_mem_32bit(
+            address,
+            data.len() as u16,
+            self.access_port.port_number(),
+        )?;
+
+        data.copy_from_slice(&received_words);
+
+        Ok(())
+    }
+
+    fn read_8(&mut self, address: u32, data: &mut [u8]) -> Result<(), ProbeRsError> {
+        self.probe.select_ap(self.access_port)?;
+
+        let received_words = self.probe.probe.read_mem_8bit(
+            address,
+            data.len() as u16,
+            self.access_port.port_number(),
+        )?;
+
+        data.copy_from_slice(&received_words);
+
+        Ok(())
+    }
+
+    fn write_word_32(&mut self, address: u32, data: u32) -> Result<(), ProbeRsError> {
+        self.probe.select_ap(self.access_port)?;
+
+        self.write_32(address, &[data])?;
+
+        Ok(())
+    }
+
+    fn write_word_8(&mut self, address: u32, data: u8) -> Result<(), ProbeRsError> {
+        self.probe.select_ap(self.access_port)?;
+
+        self.write_8(address, &[data])
+    }
+
+    fn write_32(&mut self, address: u32, data: &[u32]) -> Result<(), ProbeRsError> {
+        self.probe.select_ap(self.access_port)?;
+
+        self.probe
+            .probe
+            .write_mem_32bit(address, data, self.access_port.port_number())?;
+        Ok(())
+    }
+
+    fn write_8(&mut self, address: u32, data: &[u8]) -> Result<(), ProbeRsError> {
+        self.probe.select_ap(self.access_port)?;
+
+        let chunk_size = 64;
+
+        // If we write less than 64 bytes, just write it directly
+        if data.len() < chunk_size {
+            log::trace!("write_8: small - direct 8 bit write to {:08x}", address);
+            self.probe
+                .probe
+                .write_mem_8bit(address, data, self.access_port.port_number())?;
+        } else {
+            // Handle unaligned data in the beginning
+            let bytes_beginning = if address % 4 == 0 {
+                0
+            } else {
+                (4 - address % 4) as usize
+            };
+
+            let mut current_address = address;
+
+            if bytes_beginning > 0 {
+                log::warn!(
+                    "write_8: at_begin - unaligned write of {} bytes to address {:08x}",
+                    bytes_beginning,
+                    current_address,
+                );
+                self.probe.probe.write_mem_8bit(
+                    current_address,
+                    &data[..bytes_beginning],
+                    self.access_port.port_number(),
+                )?;
+
+                current_address += bytes_beginning as u32;
+            }
+
+            // Address has to be aligned here
+            assert!(current_address % 4 == 0);
```

Comment:
This should really be a proper error instead of a panic.

Reply:
This is just to verify that the code some lines above works. If this assert is triggered, it indicates a bug in the program, and cannot be fixed or handled by the user.